lib/cloud/gcp/vm.go

```diff
@@ -0,0 +1,545 @@
+	ProjectID string
+	// Zone is the instance's zone.
+	Zone string
+	// Name is the instance's name.
+	Name string
+}
+
+func (req *InstanceRequest) CheckAndSetDefaults() error {
+	if len(req.ProjectID) == 0 {
+		trace.BadParameter("projectID must be set")
+	}
+	if len(req.Zone) == 0 {
+		trace.BadParameter("zone must be set")
+	}
+	if len(req.Name) == 0 {
+		trace.BadParameter("name must be set")
+	}
+	return nil
+}
+
+// getHostKeys gets the SSH host keys from the VM, if available.
+func (clt *instancesClient) getHostKeys(ctx context.Context, req *InstanceRequest) ([]ssh.PublicKey, error) {
+	guestAttributes, err := clt.InstanceClient.GetGuestAttributes(ctx, &computepb.GetGuestAttributesInstanceRequest{
+		Instance:  req.Name,
+		Project:   req.ProjectID,
+		Zone:      req.Zone,
+		QueryPath: googleapi.String("hostkeys/"),
+	})
+	if err != nil {
+		return nil, trace.Wrap(convertAPIError(err))
+	}
+	items := guestAttributes.GetQueryValue().GetItems()
+	keys := make([]ssh.PublicKey, 0, len(items))
+	var errors []error
+	for _, item := range items {
+		key, _, _, _, err := ssh.ParseAuthorizedKey([]byte(fmt.Sprintf("%v %v", item.GetKey(), item.GetValue())))
+		if err == nil {
+			keys = append(keys, key)
+		} else {
+			errors = append(errors, err)
+		}
+	}
+	return keys, trace.NewAggregate(errors...)
+}
+
+// GetInstance gets a GCP VM.
+func (clt *instancesClient) GetInstance(ctx context.Context, req *InstanceRequest) (*Instance, error) {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resp, err := clt.InstanceClient.Get(ctx, &computepb.GetInstanceRequest{
+		Instance: req.Name,
+		Project:  req.ProjectID,
+		Zone:     req.Zone,
+	})
+	if err != nil {
+		return nil, trace.Wrap(convertAPIError(err))
+	}
+	inst := toInstance(resp)
+	inst.ProjectID = req.ProjectID
+
+	hostKeys, err := clt.getHostKeys(ctx, req)
+	if err == nil {
+		inst.hostKeys = hostKeys
+	} else if !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	return inst, nil
+}
+
+// SSHKeyRequest contains parameters to add/removed SSH keys from an instance.
+type SSHKeyRequest struct {
+	// Instance is the instance to add/remove keys form.
+	Instance *Instance
+	// PublicKey is the key to add. Ignored when removing a key.
+	PublicKey ssh.PublicKey
+	// Expires is the expiration time of the key. Ignored when removing a key.
+	Expires time.Time
+}
+
+func (req *SSHKeyRequest) CheckAndSetDefaults() error {
+	if req.Instance == nil {
+		return trace.BadParameter("instance not set")
+	}
+	if req.Expires.IsZero() {
+		// Default to 10 minutes to give plenty of time to install Teleport.
+		req.Expires = time.Now().Add(10 * time.Minute)
+	}
+	return nil
+}
+
+// formatSSHKey formats a public key to add to a GCP VM.
+func formatSSHKey(pubKey ssh.PublicKey, expires time.Time) string {
+	const iso8601Format = "2006-01-02T15:04:05-0700"
+	return fmt.Sprintf(`%s:%s %s google-ssh {"userName":%q,"expireOn":%q}`,
```

Comment:
Does this not require the `teleport` Linux user to be present on the instance?

Reply:
No, the `teleport` user will be created automatically by GCP.